### DISCUSSION TITLE
Auto print speeds based on filament max volumetric speed

### DISCUSSION THREAD

**Discussion by DARKFiB3R** [OrcaSlicer/OrcaSlicer#6455] [Ideas]
Is there an reason that print speeds are not automatically set or maybe just suggested after performing a max flow rate calibration?

Seems like it would be a helpful feature, or am I missing something obvious?

**Comment by kcorbett95** [comment]
I have seen this mentioned before here: https://github.com/SoftFever/OrcaSlicer/discussions/2310

I would also love to see this implemented, as it would be a really simple equation taking into account the layer height and line width for each type of line, and a percentage input for the speed settings. 

Currently I just have a spreadsheet that calculates my speeds for me based on the yellow highlighted input cells here (bottom of the sheet, used the same doc as my adaptive pressure advance calcs): https://docs.google.com/spreadsheets/d/1gnoZVrBREtT5nZ3WFdoOx85Rk9hrebyvBnhRNOkVxhg/edit?usp=sharing
It can be a bit of a pain to set up for all your print profiles, since you need to make sure the line widths for each type of line correspond to what is in the sheet for the calculation to work.

This would be super useful in maximizing print speed when using variable layer height, while maintaining maximum print quality since the printer will never try to exceed the max volumetric flowrate, but still maximize print speed for each layer and line type with the user's desired % of max flow for each line type.

**Reply by drcowsley** [reply]
1.  How is max flow rate used?  I have assumed it is used to limit print speed.
2.  How can I set a different max flow rate for a different nozzle diameter?  The relationships between nozzle diameter, flow rate, pressure and temperature is complex and depends on the material sample as much as the material type.  These are not Newtonian fluids.

**Reply by kcorbett95** [reply]
> 1. How is max flow rate used?  I have assumed it is used to limit print speed.

Currently it is used as a cap limit, for example if you set your speed to 1000mm/s but your max flowrate can only support 20mm^3/s, your printer speed will be limited well below 1000mm/s, to a value determined by your layer height and line width. 

> 2. How can I set a different max flow rate for a different nozzle diameter?  The relationships between nozzle diameter, flow rate, pressure and temperature is complex and depends on the material sample as much as the material type.  These are not Newtonian fluids.

This is already being done by the slicer in the max flow rate calculation. You would simply do the max flowrate calibration and use that number as a maximum reference for what the hotend can do with that temperature and material.

This feature would essentially allow the user to enter a percentage value for a speed, which would then refer back to the max flowrate calibration value to calculate the desired print speed. For example, I am able to push 22mm^3/s of PLA through a 0.4mm nozzle via the max flowrate cal. Picking some example values, I would then set the sparse infill to 100% and first layer to 30%. If the line width was set to 0.4mm and layer height set to 0.2mm, the slicer would automatically calculate the speeds:
sparse_speed = 100% * max_flow / (line_width * layer_height) = 1 * 22/(0.4 * 0.2) = 275 mm/s which corresponds to 22mm^3/s
layer1_speed = 30% * max_flow / (line_width * layer_height) = 0.3 * 22/(0.4 * 0.2) = 82.5 mm/s which corresponds to ~6.6mm^3/s

With this in mind, it's easy to just calculate the desired absolute speed values for a given combination of layer height and line width, but then you will run into issues with the variable layer height causing you to slam into your volumetric flow limit for all line types. For example, say you have a part that has very little in the way of vertical features, and can be printed with a much larger layer height in certain sections. If you leave your speed values set as absolutes, your outer walls (generally printed slower for more detail, I use 70% of max flow) will now hit the volumetric flow rate limit for larger layer heights since the absolute speed value was calculated at a lower layer height. In order to avoid print quality issues for parts that can take advantage of variable layer height, I usually set my print speeds using this framework according to the largest layer height, and thus the lowest absolute speeds. This means I lose out on printing time that could be sped up significantly in the lower layer height sections of the part, since I am limiting to absolute speeds even though my hotend can support significantly more flow.

Basically, allowing the slicer to calculate the absolute speed from the users input of a RELATIVE speed (as a % of max flow rate) will allow for the printer speed to be adaptive to variable layer heights, which would improve print speed for finer detail features with lower layer heights, and print quality and strength for larger layer heights, removing the tradeoff of using variable layer height at the expense of print time (lower absolute speed even for fine detail, small layer height sections where the extruder can handle a lot more flow but is not taking advantage of this) or print quality (higher absolute speed even for coarser large layer height areas where the max flowrate limit will be hit, removing the benefits of having different volumetric flowrates for the various line types) 

I don't think this would add a significant burden to development since this calculation is already being done in reverse for the max flow calibration, it would give users a way to "automatically" get rough print speeds based on a single calibration, rather than blindly relying on defaults and having to tweak every single speed value for a given combination of layer height, line width, and line type.

**Comment by drcowsley** [comment]
A 0.2mm nozzle has a quarter the hole area of a 0.4 mm nozzle, and a 0.6mm
nozzle more than twice.

For the same pressure and temperature, the maximum flow rate through a
0.6mm nozzle will be much more than twice that through a 0.4 nozzle,
assuming the same shear rate around the greater circumference and the
larger radius.

Similarly the maximum flow rate through a 0.2mm nozzle will be much less
then a quarter of that through a 0.4mm nozzle.

The best you can hope for with a 0.2 nozzle is half the layer height, half
the line width and the same speed.

All of this assumes the length of the drilled nozzle hole is constant.  If
the liquid has to go twice as far through the narrow tube it will need
something like twice the pressure to push it at the same temperature.

DrC

**Reply by dbaarda** [reply]
Note that flow viscosity limits through the nozzle are only part of the flow-rate equation. Back-pressure from the print surface also makes a big difference, which means line-width, layer-height, and head-speed have an impact. See the following were I summarised my findings so far;

https://github.com/SoftFever/OrcaSlicer/discussions/10832#discussioncomment-14603955

**Reply by kcorbett95** [reply]
Yep this all makes sense. In the case of using a smaller nozzle, when the user does their max flowrate calibration (already has been a thing in orcaslicer for years now), that would determine the reference max volumetric flowrate for that nozzle specifically, and the print speeds would be calculated as a percentage of that maximum flow. 

Another benefit for this would be to keep the cooling consistent when slicing with arachne, where areas with thinner layer width (and thus lower volumetric flow for a given layer height) would be subject to more cooling (potentially too much) relative thicker areas if printed at the same speed (not flow, toolhead velocity).

Essentially this feature would simplify tuning the print speeds for a given print profile, giving the user a smarter pre-calculated starting point for print speeds rather than absolute default values that may or may not work well at all for a given material/nozzle type/etc

**Comment by kcorbett95** [comment]
Just noticing that there is now a checkbox for adaptive volumetric speed in the latest nightly, but poking through the commits I can't find any documentation other than the bot updating the wiki with a vague description. Anyone have any idea what this new setting is doing, and/or if it relates to this?
